### PR TITLE
DIMAT Integration

### PR DESCRIPTION
**Pull request by mnpham0417**
# Description
Integrate DIMAT (Decentralized Iterative Merging-And-Training, CVPR 2024) into APPFL as a pluggable aggregator and trainer. DIMAT addresses the permutation symmetry problem in neural networks by aligning model feature spaces via activation matching before merging.

### Core Components Added

- **`DIMATaggregator`** (`src/appfl/algorithm/aggregator/dimat_aggregator.py`) — Server-side aggregator that performs activation-matching model merging using a proxy dataset. Supports ResNet (20/18/50), VGG (11/16), and CNN architectures via pluggable graph functions.
- **`DIMATTrainer`** (`src/appfl/algorithm/trainer/dimat_trainer.py`) — Client-side trainer extending `VanillaTrainer` that preserves server-computed BatchNorm statistics.
- **`dimat_utils/`** — 7 utility modules implementing the full DIMAT pipeline: computation graph construction, activation metrics (covariance/mean), feature-space alignment (Hungarian algorithm, ZipIt, optimal matching), and model merging with transformation propagation.

### Two-Phase Training

DIMAT uses a two-phase approach orchestrated by `run_mpi_dimat_pretrain.py`:
1. **Pre-training**: Each client trains independently for N epochs (no communication)
2. **Merge-train**: Standard FL loop where the server merges models using activation matching before redistributing

### Supporting Files

- Example configs for CIFAR-10, CIFAR-100, and MNIST with DIMAT
- Dataset loaders with **lazy augmentation** (fresh transforms per access) and **class-balanced IID partitioning**
- Proxy dataset loaders for server-side activation statistics
- ResNet20 (w=8) model definition matching the DIMAT paper

### Key Design Decisions

- **Lazy data augmentation**: APPFL's built-in `Dataset` class pre-computes transforms, freezing randomness. Our loaders return `Subset` wrappers around raw torchvision datasets to preserve per-access augmentation.
- **Server-side BN reset only**: The server resets BatchNorm using the full proxy dataset. The client trainer deliberately skips re-resetting with its smaller local partition.
- **Mixed-precision (autocast)**: Matches the original DIMAT code's use of `torch.amp.autocast()` during BN reset and activation computation.

### Fixes
- N/A

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (internal implementation changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (no changes to the code)
- [ ] CI change (changes to CI workflows, packages, templates, etc.)
- [ ] Version changes (changes to the package or dependency versions)

## Testing
All experiments were run on a single GPU node via MPI (1 server + N clients). DIMAT uses a two-phase setup: local pre-training followed by merge-train rounds where the server aligns and merges client models using activation matching on a proxy dataset. Standard FL baselines (FedAvg, FedAvgM, FedAdam, FedYogi, FedAdagrad) use the same total training budget (global_epochs × local_epochs matching DIMAT's pretrain + merge-train epochs), the same model, optimizer (Adam lr=0.001), and data partition for fair comparison. DIMAT's proxy dataset is 500 samples from the test set for MNIST, CIFAR-10, CIFAR-100.

### Benchmark: MNIST (5 clients, CNN)

| Algorithm | IID | Non-IID |
|---|---|---|
| **DIMAT** | **99.47%** | **96.41%** |
| FedAvg | 99.59% | 98.70% |
| FedAvgM | 99.29% | 99.07% |
| FedAdam | 99.47% | 98.35% |
| FedYogi | 99.42% | 98.58% |
| FedAdagrad | 99.37% | 90.97% |

### Benchmark: CIFAR-10 (5 clients, ResNet20)

| Algorithm | IID | Non-IID (2 cls/client) |
|---|---|---|
| **DIMAT** | **92.81%** | **25.33%** |
| FedAvg | 93.95% | 45.70% |
| FedAvgM | 91.71% | 39.99% |
| FedAdam | 92.25% | 48.33% |
| FedYogi | 92.56% | 46.90% |
| FedAdagrad | 89.29% | 42.45% |

### Benchmark: CIFAR-100 (5 clients, ResNet20)

| Algorithm | IID | Non-IID (20 cls/client) |
|---|---|---|
| DIMAT | 73.40% | -- |
| FedAvg | -- | 64.98% |
| FedAvgM | -- | 53.77% |
| FedAdam | -- | 61.73% |
| FedYogi | -- | 62.41% |
| FedAdagrad | -- | 54.73% |

**Note**: DIMAT is inherently a decentralized learning framework. Its performance and convergence rate works best under sparse communication and non-IID data distributions.

### Reproducing the Experiments

```bash
cd examples

# DIMAT (e.g., CIFAR-100 IID, 5 clients)
mpirun --oversubscribe -n 6 python mpi/run_mpi_dimat_pretrain.py \
    --server_config ../examples/resources/configs/cifar100/server_dimat_ring.yaml \
    --client_config ../examples/resources/configs/cifar100/client_dimat_ring.yaml \
    --pretrain_epochs 100
```

## Other Pull Request Checklist

Please confirm the PR meets the following requirements.
- [x] Code changes pass `pre-commit` (e.g., ruff, mypy, etc.) via `pre-commit run --all-files`.
- [x] Tests have been added to show the fix is effective or that the new feature works.
- [ ] New and existing unit tests pass locally with the changes.
- [ ] Docs have been updated and reviewed if relevant.

### CODE REVIEW

**Comment by Zilinghan** [comment]
Hi @mnpham0417, thanks for you PR! I will try to start the review as soon as I could

**Comment by mnpham0417** [comment]
> Hi @mnpham0417, thanks for you PR! I will try to start the review as soon as I could

Thank you @Zilinghan ! And awesome work on APPFL!

File: src/appfl/algorithm/trainer/dimat_trainer.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    ):
+        """Load model parameters from the server.
+
+        The server already resets BN stats using the full proxy dataset after
+        merging, so we must NOT re-reset here with only local data — that would
+        overwrite the server's higher-quality statistics and degrade accuracy.
+        """
+        super().load_parameters(params)
+
+    def _reset_bn_stats(self):
```
Comment (Zilinghan) [inline src/appfl/algorithm/trainer/dimat_trainer.py]: it seems that this function is not called elsewhere

File: run_original_dimat.py
Comment (Zilinghan) [inline run_original_dimat.py:1]: all `run_*` within the `src/appfl` folders are deprecated runners, so I will create a grpc and a serial runner within `examples/serial` and `examples/grpc` later

Comment (Zilinghan) [reply run_original_dimat.py:1]: actually, this seems to be irrelevant to appfl, could this be deleted?

**Comment by Zilinghan** [comment]
Hi @mnpham0417 , I am done with my review and created a [backward PR](https://github.com/mnpham0417/APPFL/pull/1) to your forked repo, so that you can check if my changes make sense. If so, please feel free to create a PR using that version, and we can proceed to merge, thanks!

**Comment by mnpham0417** [comment]
> Hi @mnpham0417 , I am done with my review and created a [backward PR](https://github.com/mnpham0417/APPFL/pull/1) to your forked repo, so that you can check if my changes make sense. If so, please feel free to create a PR using that version, and we can proceed to merge, thanks!

Thank you @Zilinghan! I have created [new PR](https://github.com/APPFL/APPFL/pull/389) based on your suggestion. Closing this.
